### PR TITLE
Fix footnote handling in rustdoc

### LINKED ISSUE #100638: Footnotes are squished together by rustdoc when there's no empty line between them
### Problem

Footnote lines, when written on adjacent lines, squish into a single footnote:

```rust
//! [^1]: Footnote A.
//! [^2]: Footnote B.
//! [^3]: Footnote C.
```

Screenshot 2022-08-16 at 18 39 09

### Steps

1. Copy-paste the following snippet at the top of a new module:

    ```rust
    //! Reference to footnotes A[^1], B[^2] and C[^3].
    //!
    //! [^1]: Footnote A.
    //! [^2]: Footnote B.
    //! [^3]: Footnote C.
    ```
2. Run `cargo doc`.
3. Observe that the footnotes have been incorrectly rendered as a single footnote.

### Possible Solution(s)

_No response_

### Notes

One workaround is to separate the footnotes with empty lines:

```rust
//! [^1]: Footnote A.
//! 
//! [^2]: Footnote B.
//! 
//! [^3]: Footnote C.
```

It's not pretty, though.

### Version

```text
cargo 1.64.0-nightly (b1dd22e66 2022-07-09)
release: 1.64.0-nightly
commit-hash: b1dd22e668af5279e13a071ad4b17435bd6bfa4c
commit-date: 2022-07-09
host: x86_64-apple-darwin
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.79.1 (sys:0.4.55+curl-7.83.1 system ssl:(SecureTransport) LibreSSL/3.3.6)
os: Mac OS 12.5.0 [64-bit]
```
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #100638.

You can now declare footnotes like this:

```rust
//! Reference to footnotes A[^1], B[^2] and C[^3].
//!
//! [^1]: Footnote A.
//! [^2]: Footnote B.
//! [^3]: Footnote C.
```

r? @notriddle

### CODE REVIEW

File: tests/rustdoc/footnote-definition-without-blank-line-100638.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+//! Reference to footnotes A[^1], B[^2] and C[^3].
+//!
+//! [^1]: Footnote A.
+//! [^2]: Footnote B.
+//! [^3]: Footnote C.
+
+// @has 'foo/index.html'
+// @has - '//*[@class="docblock"]/*[@class="footnotes"]/ol/li[@id="fn1"]/p' 'Footnote A'
+// @has - '//li[@id="fn1"]/p/a[@href="#fnref1"]' ''
```
Comment (fmease) [inline tests/rustdoc/footnote-definition-without-blank-line-100638.rs:11]:
```suggestion
// @has - '//li[@id="fn1"]/p/a/@href' '#fnref'
```

Might be slightly nicer but whatever

**Review by fmease — APPROVED** [review]

File: tests/rustdoc/footnote-definition-without-blank-line-100638.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply tests/rustdoc/footnote-definition-without-blank-line-100638.rs:11]: Fair enough!

**Comment by GuillaumeGomez** [comment]
Applied @fmease suggestion. :)

**Comment by fmease** [comment]
Cool!
r? fmease @bors r+ rollup

**Pull request merged**